lib/src/builder/builder.dart

```diff
@@ -1,8 +1,144 @@
+import 'dart:async';
+
+import 'package:analyzer/analyzer.dart';
 import 'package:build/build.dart';
-import 'package:source_gen/source_gen.dart';
+import 'package:dart_style/dart_style.dart';
+import 'package:over_react/src/builder/generation/declaration_parsing.dart';
+import 'package:over_react/src/builder/generation/impl_generation.dart';
+import 'package:path/path.dart' as p;
+import 'package:source_span/source_span.dart';
 
 import './util.dart';
-import './generator.dart';
 
-Builder overReactBuilder(BuilderOptions options) =>
-    new PartBuilder([new OverReactGenerator()], outputExtension);
+Builder overReactBuilder(BuilderOptions options) => new OverReactBuilder();
+
+class OverReactBuilder extends Builder {
+  static final _headerLine = '// '.padRight(77, '*');
+
+  static final _formatter = new DartFormatter();
+
+  static final RegExp _overReactPartDirective = new RegExp(
+    r'''['"].*\.over_react\.g\.dart['"]''',
+  );
+
+  static bool _mightContainDeclarations(String source) {
+    return ParsedDeclarations.key_any_annotation.hasMatch(source) ||
+      _overReactPartDirective.hasMatch(source);
+  }
+
+  @override
+  FutureOr<void> build(BuildStep buildStep) async {
+    final source = await buildStep.readAsString(buildStep.inputId);
+    if (!_mightContainDeclarations(source)) {
+      return;
+    }
+
+    CompilationUnit libraryUnit;
+    try {
+      libraryUnit = parseCompilationUnit(
+        source,
+        name: buildStep.inputId.path,
+        suppressErrors: false,
+        parseFunctionBodies: true);
+    } catch (error, stackTrace) {
+      log.severe('There was an error parsing the compilation unit for file: ${buildStep.inputId}');
+      log.fine(error);
+      log.fine(stackTrace);
+      return;
+    }
+
+    if (isPart(libraryUnit)) {
+      return;
+    }
+
+    final librarySourceFile = new SourceFile.fromString(
+      source,
+      url: idToPackageUri(buildStep.inputId));
+    final libraryDeclarations = new ParsedDeclarations(
+      libraryUnit,
+      librarySourceFile,
+      log);
+    if (libraryDeclarations.hasErrors) {
+      log.severe('There was an error parsing the file declarations for file: ${buildStep.inputId}');
+      return;
+    }
+
+    final outputs = <String>[];
+    final libraryGenerator = new ImplGenerator(log, librarySourceFile)
+      ..generate(libraryDeclarations);
+    outputs.add(libraryGenerator.outputContentsBuffer.toString());
+
+    final parts = libraryUnit.directives
+      .whereType<PartDirective>()
+      // Ignore `.over_react.g.dart` parts - that's what we're generating.
+      .where((part) => !part.uri.stringValue.endsWith(outputExtension));
```

Comment:
We may also need to ignore parts from other generated files. That was an issue with the old approach, did you verify that it's not an issue here?

Reply:
I didn't see any issues, but this reminds me that I meant to add an another check for each part directive so that we only parse the compilation units if the part file contains over react declarations. Currently this is only doing that check on the parent library file

Reply:
I think it would be safer to ignore anything ending in `.g.dart` and explicitly not support over_react annotations in generated files. We could run into a situation where we're parsing a file that doesn't exist yet.